### DISCUSSION TITLE
Context specific for each state

### DISCUSSION THREAD

**Discussion by SandroMaglione** [statelyai/xstate#3624] [Ideas]
Hi,

One of the features that xstate is missing (or at least I did not find) is the ability to have **context specific for each state**.

As a simple example, I have a file upload machine:
- `NO FILE` (initial state)
- `Load file` event that takes a `React.ChangeEvent<HTMLInputElement>`
- `FILE UPLOADED`: state reached if the file metadata (type, size, etc.) is valid

In this state machine it should be possible to specify that `NO FILE` **must** have no `File` (empty context), while `FILE UPLOADED` **must require** a `File` (`{ file: File }`).

Currently by using the "global" context it is still possible to reach `FILE UPLOADED` while somehow forgetting to assign `File`. Instead, the machine should guarantee that if the state matches `FILE UPLOADED`, then I can access a non `undefined` file.

One way to implement this behaviour is using guards. Nonetheless, I found in many similar situations that state-specific context is often present, and that using guards everywhere becomes verbose. Furthermore, it is still possible to forget adding a guard.

Is there some way to implement this behaviour that I missed? Would something like this be possible to have?

**Comment by davidkpiano** [comment]
This is something we are considering options for in v5.

It sounds like you want _state-specific context **types**_ rather than context that is specific to state nodes, which would be difficult to represent in the `State` object?

I was thinking of a design-by-contract-like approach for this (pseudocode):

```js
// tentative API - pseudocode
const fileMachine = createMachine({
  // ...
  states: {
    'no file': {
      invariant: (context): context is { file: null } => context.file === null,
      // ...
    },
    'load file': {
      // ...
    },
    'file uploaded': {
      invariant: (context): context is { file: File } => Boolean(context.file),
      // ...
    }
  }
});
```

In theory, this could also safely type each state so you get the expected `context` when using `state.matches(...)`, `state.hasTag(...)`, etc.

**Reply by SandroMaglione** [reply]
> It sounds like you want state-specific context types rather than context that is specific to state nodes

Yes, the idea would be to add the context type **directly** attached to a state. At the same time having a way to **extract (validate) the global context for each state**.

Here are some ideas supporting this option:
- Ability to add and visualize the context of each state in the editor. This would allow all the stakeholders to see what information are available at each state (a designer for example will know what text is possible to display at each point). Using `invariant` (as for your example) I think in theory this would not be possible (or is it?).
- It seems to me that `invariant: (context): context is { file: File }` is an implicit way to say "The context here requires a valid `File`". Having the context type associated to the state (and typed) instead would make this explicit.

An alternative idea would be to require each state to have `invariant` if the context type is defined. The `invariant` function should return a valid context or `null` if no valid context is found. Furthermore, the state will also require another parameter to specify in which other state to fallback if the context is invalid:

```ts
// tentative API - pseudocode
const fileMachine = createMachine({
  // ...
  states: {
    'no file': {
      context: {} as {}, // <- `invariant` not required, since no context (same as not adding `context` at all)
      // ...
    },
    'load file': {
      // ...
    },
    'file uploaded': {
      context: {} as { file: File }, // <- `invariant` required!
      fallback: 'no file', // <- fallback to `no file` when the context is invalid
      invariant: (context) => Boolean(context.file) ? { file: context.file } : null, // <- Extract valid context from the global context (return type as `{ file: File } | null`)
      // ...
    }
  }
});
```


Some other benefits for state (or state type) context:
- By using _Hierarchical states_ it would be possible to have a "sub-global" context associated to each state inside the parent (each child can access the parent context in the hierarchy)

> This is something we are considering options for in v5.

Should this be added to the [Roadmap](https://github.com/statelyai/xstate/discussions/742)?

***

I would be happy to look deeper into this to contribute and see what’s possible. Should I open an issue for this?

**Reply by davidkpiano** [reply]
Here's some things to consider:

- How should nested/parallel contexts coalesce? E.g.:
 
  ```js
  parent: {
    context: { ... },
    states: {
      child: {
        context: { ... }
      }
    }
  }
  ```

  When in the `parent.child` state, how should the contexts be combined? What if they conflict?

- What if one state specifies specific context and the next state does not? Do we _remove_ the specific context fields when transitioning to the next state?

Something that could "resolve" this, at the expense of being a bit more awkward, is to represent the contexts of each state as a keyed map:

```js
{
  context: { ... }, // global context
  value: { foo: 'bar' },
  event: { ... },
  // ...
  states: {
    foo: {/* foo specific context */},
    'foo.bar': {/* bar specific context */}
  }
}
```

So now instead of reading some value from e.g. `state.context.somewhere`, it has to either be read from `state.states['foo.bar'].somewhere`, or merged in, which can have conflicts.

This is why the idea of each state specifying its own "local" context is a bit problematic, but the idea of each state enforcing what the overall context shape should be is actually more natural. We can either treat state nodes as qualitative descriptors of the overall state (e.g. "when state is in `file uploaded`, the `ctx.file` exists and is a `File`) or we can treat state nodes as nodes on a tree (e.g. "the `file uploaded` state node, if it exists, has a `file` property of type `File` in its local context).

---

To best determine what we should do, and as a good exercise, I would appreciate a pseudocode example of what you expect the behavior to be for state-specific context, if the feature existed. The questions to answer are:

- Does each state node contribute to overall `state.context` or some specific `state.states[...]` context?
- If it contributes, how do you deal with merging/conflicts? (nested/parallel states)
- If it does not contribute, what's the best way to express each state-specific context on the overall `State` object?

**Comment by andraz** [comment]
I am solving similar problem for the purpose of JSX rendering: I would like to have the context hold a string of for example the title of JSX component conditionally rendered below the title matching the state.

The issue I am getting into is that it is not serializable as there is an `assign` function involved:

```javascript
createMachine({
      context: { title: '' },
      id: 'component',
      initial: 'initial',
      states: {
        initial: {
          entry: assign({ title: 'Welcome. You are not signed in.' }),
          on: {
            COMPANY_ADMIN: {
              target: 'company admin',
            },
            COMPANY_EMPLOYEE: {
              target: 'company employee',
            },
            PLATFORM_ADMIN: {
              target: 'platform admin',
            },
          },
        },
        'company admin': {
          entry: assign({ title: 'Company Admin' }),
          on: {
            EXIT: {
              target: 'initial',
            },
          },
        },
        'company employee': {
          entry: assign({ title: 'Company Employee' }),
          on: {
            EXIT: {
              target: 'initial',
            },
          },
        },
        'platform admin': {
          entry: assign({ title: 'Platform Admin' }),
          on: {
            EXIT: {
              target: 'initial',
            },
          },
        },
      },
    })
  )
```

Can I implement the above in raw JSON so it can be stored without functions + still execute assign to context when entering the state?

**Reply by andraz** [reply]
The JSX part for machine above:

```javascript

  return (
    <div className="flex flex-col">
      <h1>{state.context.title}</h1>
      <div className=" flex flex-1 gap-2">
        <button className="border" onClick={_ => send('COMPANY_ADMIN')}>
          company admin
        </button>
        <button className="border" onClick={_ => send('COMPANY_EMPLOYEE')}>
          company employee
        </button>
        <button className="border" onClick={_ => send('PLATFORM_ADMIN')}>
          platform admin
        </button>
        <button className="border" onClick={_ => send('EXIT')}>
          exit
        </button>
      </div>
      {state.matches('platform admin') && (
        <Suspense fallback={<div>Loading...</div>}>
          <E>
            <Admin GenericAdmin={GenericAdmin} />
          </E>
        </Suspense>
      )}
      {(state.matches('company admin') ||
        state.matches('company employee')) && (
        <Suspense fallback={<div>Loading...</div>}>
          <E>
            <Dashboard />
          </E>
        </Suspense>
      )}
    </div>
  )
```
